File: src/Java/ArraysSort/Base/MergeSort.java
/**
 * Acwing.787 逆序对的数量 https://www.acwing.com/problem/content/789/
 */
package Java.ArraysSort.Base;

import java.io.*;

public class MergeSort {
    static void sort(int[] arr, int begin, int l1, int l2, int[] array) {
        int i = begin, k = 0, j = begin + l1;
        while (i != begin + l1 && j != begin + l1 + l2) {
            if (arr[j] >= arr[i]) {
                array[k++] = arr[i++];
            }
            if (arr[j] < arr[i]) {
                array[k++] = arr[j++];
            }
        }

        while (j != begin + l1 + l2) {
            array[k++] = arr[j++];
        }

        while (i != begin + l1) {
            array[k++] = arr[i++];
        }

        for (int l = 0; l < l1 + l2;) {
            arr[l + begin] = array[l++];
        }
    }

    static void mergeSort(int[] arr, int begin, int end, int array[]) {
        if (begin >= end) {
            return;
        }
        int mid = (begin + end) / 2;
        mergeSort(arr, begin, mid, array);
        mergeSort(arr, mid + 1, end, array);
        sort(arr, begin, mid - begin + 1, end - mid, array);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(in.readLine());
        int[] arr = new int[n];
        int[] array = new int[n];

        String numbers = in.readLine();
        if (numbers == null) {
            return;
        }
        String[] strNumbers = numbers.split(" ");
        for (int i = 0; i < strNumbers.length; i++) {
            arr[i] = Integer.parseInt(strNumbers[i]);
        }
        in.close();

        mergeSort(arr, 0, n - 1, array);

        BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out));
        out.write(Integer.toString(arr[0]));
        for (int i = 1; i < n; i++) {
            out.write(" " + arr[i]);
        }
        out.newLine();
        out.flush();
        out.close();
    }
}

File: Python/Sort/MergeSort.py
# AcWing.787

def sort_list(array: list, start: int, l1: int, l2: int):
    i = start
    j = start + l1
    new_array = []
    while i < start + l1 and j < start + l1 + l2:
        while array[j] > array[i] and i < start + l1:
            new_array.append(array[i])
            i = i + 1
        new_array.append(array[j])
        j = j + 1
    if i == start + l1:
        while j < start + l1 + l2:
            new_array.append(array[j])
            j = j + 1
    else:
        while i < start + l1:
            new_array.append(array[i])
            i = i + 1
    array[start: start + l1 + l2] = new_array


def merge(array: list, begin: int, end: int):
    if begin >= end:
        return
    mid = int((begin + end) / 2)
    merge(array, begin, mid)
    merge(array, mid + 1, end)
    sort_list(array, begin, mid - begin + 1, end - mid)


num = int(input())
numbers = list(map(int, input().split()))
merge(numbers, 0, num - 1)
for x in numbers:
    print(x, end=' ')
